### PR TITLE
Specify form refactoring

### PR DESCRIPTION
**Pull request by peterdavidhamilton**
- fix a bug caused by a FaF request not being submitted
- fix bug when guests swap to search for an org of a different type group vs. school
- fix missing org confirmation when editing
- refactor form logic for improved unit testing
- infer form values from the respondent, passed as a param
- tidy code in views, controllers and forms
- add missing coverage
- combine `school_urn` and `group_urn` into `org_id`, including reversible data migration

# Next Steps

Continue refactor:

- create service to be single point of contact between Specify and Supported org data queries https://trello.com/c/Me3nKtOx/394-use-of-supported-gias-data-in-specify 
- common struct interface abstracting distinction between school and group

### CODE REVIEW

File: app/forms/framework_support_form.rb [hunk 1]
```diff
@@ -12,87 +12,172 @@
...
+  def login_with_many_orgs?
+    !user.guest? && !user.single_org?
+  end
+
+  # @return [Boolean]
+  def login_with_inferred_org?
+    !user.guest? && user.single_org?
+  end
+
+  # @return [Boolean]
```
Comment (peterdavidhamilton) [inline app/forms/framework_support_form.rb:173]: Or `nil`

**Comment by peterdavidhamilton** [comment]
Current output but improvements can be made:

```
Edit an unsubmitted 'Find a Framework' request
  when user is a guest
        allows the user to enter their email    
        allows the user to enter their name    
        allows the user to change their message    
        allows the user to change their selected school    
  when user is already logged in
    when user with many supported schools
          allows the user to change the school    
    when user with one supported school
          does not allow to change the school    

Submit a 'Find a Framework' request
      confirms where the message was sent and the response time    

'Find a Framework' journey
      links to external resources in a new tab    
      links to the homepage    
      loads without requiring authentication    
  for authenticated users
        shows their profile    
    when user signs out
          redirects to the start page    

'Find a Framework' request
  when not signed in
    log in choice page
          has a back link to the start page    
          errors if no selection is made    
          asks if a DSI account is available    
      and user selects DfE Sign-in
            redirects to DSI    
  when signed in
    and organisation can be inferred
      message page
            is the only step    
            raises a validation error if missing    
            goes back to the start page    
    and organisation cannot be inferred
          errors if a school is not selected    
          lists all supported schools    
          proceeds to the message page    

'Find a Framework' request as a guest
  when choosing a group or trust
    the search for a group page
          validates group field    
          goes back to choosing school or group    
          has the correct attributes    
          links to the search for a school page    
    the group confirmation page
          displays a validation message if no selection made    
          goes back to the search for a group page    
          has the correct attributes    
  what type of organisation page
        loads the page    
        errors if no selection is given    
        has a back link to the 'Do you have a DfE Sign-in' page    
  when selecting a single school
    the name page
          has the correct attributes    
          goes back to the school search page    
          raises a validation error when nothing entered    
    the message text page
          has the correct attributes    
          goes back to the email address page    
          validates the help message field    
    the email address page
          has the correct attributes    
          validates email field    
          goes back to the name page    
    the school search page
          validates school field    
          has a link to the group search page    
          has the correct attributes    
          goes back to the 'What type of organisation are you buying for?' page    
    the school details page
          choosing no goes back to the school search page    
          goes back to the school search page    
          has the correct attributes    

'Find a Framework' answers
  when completed as an authenticated user
        only allows some fields to be edited    
        has submission information    
        goes back to the message page    
  when completed as a guest
        goes back to the message page    
        allows all fields to be edited    
        has submission information    

```

**Comment by peterdavidhamilton** [comment]
[Is this the school you're buying for? page; selecting the 'continue' button allows user to navigate to the next page without selecting an option](https://trello.com/c/s8PG8jLF/329-is-this-the-school-youre-buying-for-page-selecting-the-continue-button-allows-user-to-navigate-to-the-next-page-without-selectin)

File: spec/forms/self-serve/framework_support_form_spec.rb [hunk 2]
```diff
@@ -1,92 +1,100 @@
...
+      let(:user) { create(:user, :many_supported_schools_and_groups) }
 
-        it "clears the group value" do
-          form.forget_org
-          expect(form.to_h).to eql({
-            school_urn: "1234",
-            group_uid: nil,
-          })
-        end
+      it "message > oerg select" do
```
Comment (EdwinKruglov) [inline spec/forms/self-serve/framework_support_form_spec.rb:87]: Tiny typo.

**Comment by peterdavidhamilton** [comment]
```
Submit a 'Find a Framework' request
  confirms where the message was sent and the response time

'Find a Framework' answers
  when completed as a guest
    has submission information
    goes back to the message page
    allows all fields to be edited
  when completed as an authenticated user
    has submission information
    goes back to the message page
    only allows some fields to be edited

'Find a Framework' request as a guest
  when selecting a single school
    the message text page
      goes back to the email address page
      validates the message is entered
      has the correct attributes
    the school search page
      validates the selected school
      has the correct attributes
      goes back to the organisation choice page
      has a link to the group search page
    the name page
      goes back to the school search page
      asks for their full name
      validates their name is entered
    the school confirmation page
      goes back to the school search page
      choosing no goes back to the school search page
      has the correct attributes
    the email address page
      has the correct attributes
      validates their email is entered
      goes back to the name page
  the organisation type choice page
    errors if no selection is given
    asks for school or group
    goes back to the login choice page
  when choosing a group or trust
    the search for a group page
      goes back to choosing school or group
      validates the selected group
      links to the search for a school page
      asks to search for a group
    the group confirmation page
      validates the group is confirmed
      choosing no goes back to the group search page
      has the correct attributes
      goes back to the group search page

'Find a Framework' request
  when not signed in
    the log in choice page
      errors if no selection is made
      has a back link to the start page
      asks if a DSI account is available
      and user selects DfE Sign-in
        redirects to DSI
  when signed in
    and organisation can be inferred
      the message page
        validates the message exists
        is the only step
        goes back to the start page
    and organisation cannot be inferred
      lists all supported schools and groups
      validates a group or school is selected
      proceeds to the message page

'Find a Framework' journey
  links to the homepage
  loads without requiring authentication
  links to external resources in a new tab
  for authenticated users
    shows their profile
    when user signs out
      redirects to the start page

Edit an unsubmitted 'Find a Framework' request
  when user is a guest
    allows the user to enter their email
    allows the user to change their message
    allows the user to enter their name
    allows the user to change their selected school
  when user is already logged in
    when user with one supported school
      does not allow to change the school
    when user with many supported schools
      allows the user to change the school
```

File: spec/features/self-serve/faf/answers_page_spec.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
+RSpec.feature "'Find a Framework' answers" do
+  let(:keys) { all("dt.govuk-summary-list__key") }
+  let(:values) { all("dd.govuk-summary-list__value") }
+  let(:actions) { all("dd.govuk-summary-list__actions") }
+
+  context "when completed as an authenticated user" do
```
Comment (shafiqahmedkhan) [inline spec/features/self-serve/faf/answers_page_spec.rb:6]: @peterdavidhamilton when what is completed as an authenticated user? can we be more specific.

Comment (peterdavidhamilton) [reply spec/features/self-serve/faf/answers_page_spec.rb:6]: To achieve that syntax might be hard sometimes. The "what" is "the answers to a FaF request"
Is this acceptable?

- The 'Find a Framework' check your answers page, when completed as a guest, allows all fields to be edited
- The 'Find a Framework' check your answers page, when completed as an authenticated user, only allows some fields to be edited

Comment (peterdavidhamilton) [reply spec/features/self-serve/faf/answers_page_spec.rb:6]: @EdwinKruglov @serenaabbott11 suggestions welcome

**Review by PiperS52 — COMMENTED** [review]
Looks good - messaged with details of a few minor bugs after pulling and running the code

**Review by shafiqahmedkhan — APPROVED** [review]

**Review by AndrewCowle — APPROVED** [review]

**Comment by peterdavidhamilton** [comment]
```
Starting a 'Find a Framework' request
  loads without requiring authentication
  links to the homepage
  links to external resources in a new tab
  for authenticated users
    shows their profile
    when user signs out
      redirects to the start page

Creating a 'Find a Framework' request as a guest
  the organisation type choice page
    goes back to the login choice page
    validates a choice is made
    asks for school or group
  when selecting a group or trust
    the search for a group page
      links to the school search page
      goes back to the organisation choice page
      has the correct attributes
      validates the selected group
    the group confirmation page
      choosing yes continues to the name page
      validates the group is confirmed
      has the correct attributes
      choosing no goes back to the group search page
      goes back to the group search page
  when selecting a single school
    the email address page
      has the correct attributes
      goes back to the name page
      validates their email
    the message text page
      goes back to the email address page
      validates the message is entered
      has the correct attributes
    the school search page
      goes back to the organisation choice page
      has the correct attributes
      links to the group search page
      validates the selected school
    the name page
      goes back to the school confirmation page
      asks for their full name
      validates their name
    the school confirmation page
      has the correct attributes
      goes back to the school search page and retains the autocompleted result
      choosing yes continues to the name page
      choosing no goes back to the school search page

Editing a 'Find a Framework' request as a guest
  goes back to the message page
  allows all fields to be edited
  edit message
  edit email
  has submission information
  edit name
  change organisation (same type)
    goes to the school search page
    the confirmation page goes back to the school search page
    when confirmed
      saves selected organisation
    when cancelled
      remains unchanged
  change organisation type and reselect
    edit type
    and reselect organisation
      when confirmed
        saves selected organisation
      when cancelled
        remains unchanged
        
Creating a 'Find a Framework' request
  when not signed in
    the log in choice page
      has a back link to the start page
      validates a choice is made
      asks if a DSI account is available
      and user selects DfE Sign-in
        redirects to DSI
  when signed in
    and organisation can be inferred
      the message page
        goes back to the start page
        validates the message exists
        is the only step
    and organisation cannot be inferred
      proceeds to the message page
      lists all supported schools and groups
      validates a group or school is selected
      when the school or group needs to be changed
        goes back so a different school or group can be selected

Editing a 'Find a Framework' request as a user
  goes back to the message page
  has submission information
  edit message
  only allows some fields to be edited
  with one supported school
    the school can not be edited
  with many supported schools and groups
    the school or group can be edited

Submitting a 'Find a Framework' request
  confirms where the message was sent and the response time
```

**Comment by peterdavidhamilton** [comment]
@d-a-v-e I am proposing changes to the FaF code that make into the sort of thing I had envisaged. Essentially all param manipulation and stepping logic in the form. In order to simplify the code I saw the need to combine the `school_urn` and `group_uid` fields into `org_id`. Data migration is included in the last commi. I know you are familiar with this area of the app so I want to get your thoughts.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -20,7 +20,7 @@
   # Screen-captures
   - Record only the browser window
   - Don't full screen the browser window (to avoid large files)
-  - Break into seperate videos if there are several journeys being preseneted
+  - Break into separate videos if there are several journeys being presented
```
Comment (d-a-v-e) [inline .github/pull_request_template.md:23]: thakns 🥴

File: app/controllers/support_requests_controller.rb [hunk 5]
```diff
@@ -2,67 +2,56 @@
...
-      advance_form
     end
   end
 
-  # all questions
   def update
-    @support_form = form
-
     if validation.success?
+      @form.toggle
```
Comment (d-a-v-e) [inline app/controllers/support_requests_controller.rb:48]: it's not immediately obvious to me what toggle does (by looking at the method). Is there a better name for this?

**Review by d-a-v-e — APPROVED** [review]
amazing refactor!!

File: app/controllers/support_requests_controller.rb (hunk 5 above)
Comment (peterdavidhamilton) [reply app/controllers/support_requests_controller.rb:48]: absolutely!

**Pull request merged**